### PR TITLE
Use `openApi` field in kustomization file to specify OpenAPI schema version

### LINKED ISSUE #3198: Support multiple static kubernetes OpenAPI schema versions
A kustomization.yaml file should be able to specify one of 5 builtin OpenAPI schema versions via an `openApi` field
```
openApi:
    version: v1.17.1
```

The `kustomize openapi info` command should report the multiple versions that are builtin. 

`kustomize build` should fail if the specified version isn't built in. 

Additionally, the process for acquiring new OpenAPI data (#2862) will have to be updated slightly to work with the new directory structure required to have multiple versions.

### PR DESCRIPTION
**Pull request by natasha41575**
fix #3198

### CODE REVIEW

**Comment by natasha41575** [comment]
@monopole where do I put documentation for this new field?

**Comment by Shell32-Natsu** [comment]
It should be https://github.com/kubernetes-sigs/cli-experimental now.

**Comment by khrisrichardson** [comment]
Any thoughts on updating references from `openApi` to `openAPI` for consistency with other code in the kustomize repo, Go naming conventions, and how it is referred at openapis.org?

**Comment by natasha41575** [comment]
@khrisrichardson I'll update the PR, that's a good suggestion. Thank you

File: kyaml/openapi/openapi.go [hunk 1]
```diff
@@ -419,29 +423,79 @@ const (
...
+	if kubernetesOpenAPIVersion != "" && !reset {
+		return nil
+	}
+
+	kubernetesOpenAPIVersion = strings.Replace(version["version"], ".", "", -1)
+	if kubernetesOpenAPIVersion == "" {
+		return nil
+	}
+	if _, ok := kubernetesapi.OpenAPIMustAsset[kubernetesOpenAPIVersion]; !ok {
+		return fmt.Errorf("the specified OpenAPI version is not built into kustomize")
```
Comment (Shell32-Natsu) [inline kyaml/openapi/openapi.go:438]: kyaml is not only used by kustomize.

File: kyaml/openapi/openapi.go [hunk 2]
```diff
@@ -419,29 +423,79 @@ const (
...
 )
 
+// SetSchemaVersion sets the kubernetes OpenAPI schema version to use
+func SetSchemaVersion(version map[string]string, reset bool) error {
+	// this should only be set once
+	if kubernetesOpenAPIVersion != "" && !reset {
+		return nil
+	}
+
+	kubernetesOpenAPIVersion = strings.Replace(version["version"], ".", "", -1)
```
Comment (Shell32-Natsu) [inline kyaml/openapi/openapi.go:433]: `version["version"]` looks weird.

File: kyaml/openapi/openapi.go [hunk 3]
```diff
@@ -389,7 +393,7 @@ func (rs *ResourceSchema) PatchStrategyAndKey() (string, string) {
 const (
 	// kubernetesAPIDefaultVersion is the latest version number of the statically compiled in
 	// OpenAPI schema for kubernetes built-in types
-	kubernetesAPIDefaultVersion = kubernetesapi.DefaultOpenApi
+	kubernetesAPIDefaultVersion = kubernetesapi.DefaultOpenAPI
```
Comment (Shell32-Natsu) [inline kyaml/openapi/openapi.go:396]: The name should keep consistent with `kubernetesOpenAPIVersion`

File: kyaml/openapi/openapi.go [hunk 4]
```diff
@@ -30,6 +33,7 @@ type openapiData struct {
 	schemaByResourceType           map[yaml.TypeMeta]*spec.Schema
 	namespaceabilityByResourceType map[yaml.TypeMeta]bool
 	noUseBuiltInSchema             bool
+	olderVersion                   bool
```
Comment (Shell32-Natsu) [inline kyaml/openapi/openapi.go:36]: The name `olderVersion` is pretty confusing. It looks like a version string. Maybe `isOlderVersion` or something.

File: api/internal/target/kusttarget.go [hunk 5]
```diff
@@ -382,6 +383,11 @@ func (kt *KustTarget) accumulateDirectory(
 		return nil, errors.Wrapf(
 			err, "couldn't make target for path '%s'", ldr.Root())
 	}
+	err = openapi.SetSchemaVersion(subKt.Kustomization().OpenAPI, false)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err, "couldn't set openapi version for path '%s'", ldr.Root())
+	}
 	if isComponent && subKt.kustomization.Kind != types.ComponentKind {
```
Comment (Shell32-Natsu) [inline api/internal/target/kusttarget.go:391]: What happens to a component with openAPI field? Need a test to cover it as well.

**Comment by monopole** [comment]
/ok-to-test

Eager to get this in.

File: kyaml/openapi/openapi.go (hunk 4 above)
Comment (monopole) [reply kyaml/openapi/openapi.go:36]: right, begs the question "older than what?"  

all the fields need explanatory comments

File: api/internal/target/kusttarget.go (hunk 5 above)
Comment (natasha41575) [reply api/internal/target/kusttarget.go:391]: Added a test for component

File: kyaml/openapi/openapi.go (hunk 1 above)
Comment (natasha41575) [reply kyaml/openapi/openapi.go]: Fixed

File: kyaml/openapi/openapi.go (hunk 3 above)
Comment (natasha41575) [reply kyaml/openapi/openapi.go]: changed to kubernetesOpenAPIDefaultVersion

File: kyaml/openapi/openapi.go (hunk 4 above)
Comment (natasha41575) [reply kyaml/openapi/openapi.go]: updated and commented these fields

[the remark this answers is not in the document]
Comment (natasha41575) [reply kyaml/openapi/openapi.go]: done

File: kyaml/openapi/openapi.go (hunk 2 above)
Comment (natasha41575) [reply kyaml/openapi/openapi.go]: changed

**Comment by natasha41575** [comment]
@monopole does this look good to go in?

**Pull request merged**